src/id_proofs.rs

```diff
@@ -0,0 +1,234 @@
+pub type VersionedIdentityProof = Versioned<IdentityProof>;
+
+impl
+    TryFrom<
+        concordium_base::common::Versioned<
+            concordium_base::id::id_proof_types::Proof<ArCurve, AttributeKind>,
+        >,
+    > for VersionedIdentityProof
+{
+    type Error = serde_json::Error;
+
+    fn try_from(
+        value: concordium_base::common::Versioned<
+            concordium_base::id::id_proof_types::Proof<ArCurve, AttributeKind>,
+        >,
+    ) -> Result<Self, Self::Error> {
+        serde_json::to_string(&value).and_then(|s| serde_json::from_str(&s))
+    }
+}
+
+/// Prove a [`IdentityStatement`] given the provided context, producing a [`VersionedIdentityProof`]
+#[allow(clippy::too_many_arguments)]
+pub fn prove_identity_statement(
+    seed: Bytes,
+    net: Network,
+    global_context: GlobalContext,
+    ip_index: u32,
+    identity_index: u32,
+    credential_index: u8,
+    identity_object: IdentityObject,
+    statement: IdentityStatement,
+    challenge: Bytes,
+) -> Result<VersionedIdentityProof, ConcordiumWalletCryptoError> {
+    let fn_name = "prove_id_statement";
+
+    let wallet = get_wallet(hex::encode(seed), net.into())
+        .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+    let global_context =
+        concordium_base::id::types::GlobalContext::<ArCurve>::try_from(global_context)
+            .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+    let id_object =
+        IdentityObjectV1::<IpPairing, ArCurve, AttributeKind>::try_from(identity_object)
+            .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+
+    let credential_context = CredentialContext {
+        wallet,
+        identity_provider_index: ip_index.into(),
+        identity_index,
+        credential_index,
+    };
+    let cred_id = credential_context
+        .wallet
+        .get_prf_key(ip_index, identity_index)
+        .context("Failed to get PRF key")
+        .and_then(|key| {
+            key.prf(&global_context.on_chain_commitment_key.g, credential_index)
+                .context("Failed to compute PRF")
+        })
+        .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+
+    let statement =
+        concordium_base::id::id_proof_types::Statement::<ArCurve, AttributeKind>::try_from(
+            statement,
+        )
+        .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+    let statement = StatementWithContext {
+        statement,
+        credential: cred_id,
+    };
+
+    let proof = statement
+        .prove(
+            ProofVersion::Version2,
+            &global_context,
+            challenge.as_ref(),
+            &id_object.alist,
+            &credential_context,
+        )
+        .context("Could not produce proof.")
+        .map_err(|e| e.to_call_failed(fn_name.to_string()))?;
+    VersionedIdentityProof::try_from(concordium_base::common::Versioned::new(VERSION_0, proof))
```

Comment:
I don't fully understand the purpose of this `VERSION_0`. Will it always be 0 for now or is it related to the `proofV2` ?

Reply:
In the JS SDK, the proof produced for `VERSION_0` is `ProofVersion::Version1`, so the version should (probably) be bumped. The proof version is needed to verify the proof.. I can't find any documentation pointing to a link between a `Versioned<Proof>` and `ProofVersion` 🤷 . Bare in mind this is for the "old" ID proofs and is not used for ID2.5 verifiable credentials so it probably will not be updated.

Have you worked with ID proofs versioned this way?

Reply:
`VERSION_0` for `ProofVersion::Version1`
and
`VERSION_1` for `ProofVersion::Version2`
makes sense, so I am fine if we go with that.